compiler/optimizer/abstractinterpreter/IDT.hpp

```diff
@@ -0,0 +1,151 @@
+ *
+ * The parent-child relationship in the IDT corresponds to the caller-callee relationship.
+ */
+class IDT
+   {
+   public:
+   IDT(TR::Region& region, TR_CallTarget*, TR::ResolvedMethodSymbol* symbol, uint32_t budget, TR::Compilation* comp);
+
+   TR::IDTNode* getRoot() { return _root; }
+
+   TR::Region& getRegion() { return _region; }
+
+   void addCost(uint32_t cost) { _totalCost += cost; }
+   uint32_t getTotalCost() { return _totalCost; }
+
+   /**
+    * @brief Get the total number of nodes in this IDT.
+    *
+    * @return the total number of node
+    */
+   uint32_t getNumNodes() { return _nextIdx + 1; }
+
+   /**
+    * @brief Get the next avaible IDTNode index.
+    *
+    * @return the next index
+    */
+   int32_t getNextGlobalIDTNodeIndex() { return _nextIdx; }
+
+   /**
+    * @brief Increase the next available IDTNode index by 1.
+    * This should only be called when successfully adding an IDTNode to the IDT
+    */
+   void increaseGlobalIDTNodeIndex()  { _nextIdx ++; }
+
+   /**
+    * @brief Get the IDTNode using index.
+    * Before using this method for accessing IDTNode, flattenIDT() must be called.
+    *
+    * @return the IDT node
+    */
+   TR::IDTNode *getNodeByGlobalIndex(int32_t index);
+
+   /**
+    * @brief Flatten all the IDTNodes into a list.
+    */
+   void flattenIDT();
+
+   void print();
+
+   private:
+   TR::Compilation* comp() { return _comp; }
+
+   TR::Compilation *_comp;
+   TR::Region&  _region;
+   int32_t _nextIdx;
+   uint32_t _totalCost;
+   TR::IDTNode* _root;
+   TR::IDTNode** _indices;
+   };
+
+/**
+ * Accessing IDTNode by the priority of its cost and benefit.
+ * This priority queue is to prepare an ordered list of inlining options as described in following page:
+ * https://patents.google.com/patent/US10055210B2/en
+ *
+ * It has 3 private variables:
+ * _idt: same as the input IDT.
+ * _entries: a vector used to store the result IDTNodes which were popped out from _pQueue.
+ * _pQueue: a priority queue keeps IDTNode in the order of priority from highest to lowest.
+ *
+ * The queue has overwrite the operator for comparing he IDTNodes:
+ * - the node with the lowest benefit will have the highest priority;
+ * - the queue breaks the tie by comparing the cost (node with lowest cost will have higher priority).
+ *
+ * Right after the priority queue is constructed, only the root of given IDT will be pushed into _pQueue.
+ * When getting the IDTNode at the specific index at the priority queue, it will check if the node has
+ * already been saved in _entries first. If yes, return the IDTNode at the given index from _entries, else:
+ *    1. pop the IDTNode which has the highest priority from _pQueue;
+ *    2. append the popped node at the end of _entries;
+ *    3. push all children of the popped node into _pQueue;
+ *    4. return the IDTNode at the given index from _entries.
+ *
+ * This priority queue traverses all predecessors in preorder before accessing a successor.
```

Comment:
Thank you for the thorough explanation of the difference among preorder traversal, reverse postorder traversal and topological sort.
Actually, this part of code has nothing about reverse postorder traversal. I guess the prior contributor mentioned reverse postorder traversal is because this traversal was mentioned in our paper, but it is actually used as a code block iterator in another part of code at our OpenJ9 contribution for phase 3.
And yes, this `IDTPriorityQueue` is actually a topological sort. I guess the only reason why the author called it as preorder is the IDT nodes are traversed from root to leaves.
I have removed redundant part of the comments and kept the link of the patent and quote from the patent.

Reply:
Thanks!